### ISSUE TITLE
Test failure of test_ace and test_all_2

### ISSUE THREAD

**Issue by aszlig** [openpaperwork/libpillowfight#2]
I'm currently packaging Paperwork for NixOS and stumbled on two failing test cases:

```
======================================================================
FAIL: test_ace (tests.tests_ace.TestACE)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/nix-build-python3.5-pypillowfight-0.2.1.drv-0/libpillowfight-0.2.1-src/tests/tests_ace.py", line 23, in test_ace
    self.assertEqual(out_img.tobytes(), expected_img.tobytes())
AssertionError: b'\xd[78253258 chars]cf\xe9\xe6\xe1\xe7\xe6\xe1\xea\xe7\xe2\xe8\xe7[405912 chars]\x00' != b'\xd[78253258 chars]cf\xe7\xe6\xe2\xe7\xe6\xe1\xe8\xe7\xe3\xe8\xe7[405912 chars]\xe8'

======================================================================
FAIL: test_all_2 (tests.tests_all.TestAll)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/nix-build-python3.5-pypillowfight-0.2.1.drv-0/libpillowfight-0.2.1-src/tests/tests_all.py", line 61, in test_all_2
    self.assertEqual(out_img.tobytes(), expected_img.tobytes())
AssertionError: b'\xf[77489438 chars]ff\xfd\xff\xff\xff\xff\xff\xfd\xff\xff\xff\xff[1169842 chars]\x00' != b'\xf[77489438 chars]ff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff[1169842 chars]\xff'
```

The Python dependencies of the closure:

```
/nix/store/acl4cx559rgmqjqs3x37zpr50lgmcam7-python3-3.6.0rc1.drv
/nix/store/aj8ry51h4ibjyqyiv4v9ns73qzz8rgvd-python3.6-bootstrapped-pip-9.0.1.drv
/nix/store/p5g9bnqwafx3wri26nqdygvjkq0d6si7-python3.6-setuptools-30.2.0.drv
/nix/store/kmv5dpffkz982ycmfamwx76jhdvhpnvy-python3.6-pbr-1.8.1.drv
/nix/store/dvbfqg2dc4wnqidd29fdk9yi5nj4ymv8-python3.6-linecache2-1.0.0.drv
/nix/store/1f9pmvf8qmk1h206732rlvrcfpsh375c-python3.6-traceback2-1.4.0.drv
/nix/store/7rd3wgym91ldby8nmwqhd02rx5h11sd8-python3.6-py-1.4.31.drv
/nix/store/yg6rpwfkrpsx6979d8jki9q0vl0f73k1-python3.6-pytest-2.9.2.drv
/nix/store/4h7z49hv70p8w95ibrgr550bds9yhfy8-python3.6-six-1.10.0.drv
/nix/store/pgynbiljbnjlmjaivwmhjia9glwqsfsq-python3.6-argparse-1.4.0.drv
/nix/store/1zh1y4vwmbnplv0f7391g1vw1axkzgd7-python3.6-unittest2-1.1.0.drv
/nix/store/50p852lclvgabph8nrz1ndxa74pxni9s-python3.6-funcsigs-1.0.2.drv
/nix/store/m2qrn6hbwpmp5n3k6wzs8rlj5rvpyifh-python3.6-mock-2.0.0.drv
/nix/store/p2n117mzmnc87yvsfc5jr05hga8x8qdz-python3.6-coverage-4.0.1.drv
/nix/store/1m1jj06qcx305zp4lf4fdwq03pf8v46x-python3.6-nose-1.3.7.drv
/nix/store/1gkv88zbrcbyb028379my9djwy2sb0pp-python3.6-Pillow-3.4.2.drv
/nix/store/g4pvijy5048wmk1za1g2crhc0drigw4x-python3.6-pypillowfight-0.2.1.drv
```

**Comment by jflesch** [comment]
That's weird. Tests are working on my system (Debian stable + Python 3.4 / Debian sid + Python 3.5).

Unfortunately, this problem can come from various things:
- Pillow: Depending on the version of Pillow, decoding of JPEG may vary slightly
- Python random generator (used by ACE): Depending on the version Python, the random generator may be slightly different and produce slightly different numbers (even with a defined seed, as done in the tests)

Basically, the reference platform for the tests is Debian stable. I cannot really support other platforms with these kind of tests :/

Regarding the name, I guess I should consider renaming it 'pypillowfight' or something else, since it seems the ReviewBoard's guys use the same package name.

**Comment by aszlig** [comment]
And another one (test_swt2) for `i686-linux`: https://hydra.nixos.org/build/45122757/nixlog/6/raw
Hm, is it possible in some way to make these tests more deterministic?

**Comment by jflesch** [comment]
I wonder .. are you working on i686 everytime, or do you work with amd64 systems too ?

**Comment by aszlig** [comment]
Nah, this was packaging for NixOS and both `i686-linux` and `x86_64-linux` are supported architectures. I for myself do not have a `i686-linux` anymore.

**Comment by QuLogic** [comment]
On x86_64, [`test_all_1` sometimes fails](https://kojipkgs.fedoraproject.org//work/tasks/5440/23925440/build.log) or [sometimes `test_ace` and `test_all_2` fails](https://kojipkgs.fedoraproject.org//work/tasks/5109/23925109/build.log). On ppc64le, it seems to [fail on `test_swt`, `test_swt2`, and `test_canny`](https://kojipkgs.fedoraproject.org//work/tasks/5441/23925441/build.log). On i386, [`test_ace`, `test_swt_2`, `test_all_2` all fail](https://kojipkgs.fedoraproject.org//work/tasks/5441/23925441/build.log).

Maybe you can import this [assert_image_similar](https://github.com/python-pillow/Pillow/blob/master/Tests/helper.py#L124) that Pillow uses?

**Comment by QuLogic** [comment]
If I hard-code 9+ instead of `cpu_count`, I can get `test_ace` to consistently fail. Looking at the failing image, it appears that the last row is black instead of the white from the receipt.

**Comment by jflesch** [comment]
@aszlig With some luck, @QuLogic  has solved your problem :)
